HpToolsLauncher/Launcher.cs

```diff
@@ -936,515 +936,520 @@ private IAssetRunner CreateRunner(TestStorageType runType, JavaProperties ciPara
-                                }
-                            }
-
-                            //Proxy password
-                            if (_ciParams.ContainsKey("MobileProxySetting_PasswordBasicAuth"))
-                            {
-                                // base64 decode
-                                byte[] data = Convert.FromBase64String(_ciParams["MobileProxySetting_Password"]);
-                                mcConnectionInfo.MobileProxySetting_Password = Encoding.Default.GetString(data);
-                            }
-                            else if (_ciParams.ContainsKey("MobileProxySetting_Password"))
-                            {
-                                string proxyPassword = _ciParams["MobileProxySetting_Password"];
-                                if (!string.IsNullOrEmpty(proxyPassword))
-                                {
-                                    mcConnectionInfo.MobileProxySetting_Password = Decrypt(proxyPassword, _secretKey);
-                                }
-                            }
-
-                        }
-                    }
-
-                    // other mobile info
-                    string mobileinfo = string.Empty;
-                    if (_ciParams.ContainsKey("mobileinfo"))
-                    {
-                        mobileinfo = _ciParams["mobileinfo"];
-                    }
-
-                    Dictionary<string, List<String>> parallelRunnerEnvironments = new Dictionary<string, List<string>>();
-
-                    // retrieve the parallel runner environment for each test
-                    if (_ciParams.ContainsKey("parallelRunnerMode"))
-                    {
-                        if (Convert.ToBoolean(_ciParams["parallelRunnerMode"]))
-                        {
-
-                            foreach (var test in validTests)
-                            {
-                                string envKey = "Parallel" + test.Id + "Env";
-                                List<string> testEnvironments = GetParamsWithPrefix(envKey);
-
-                                // add the environments for all the valid tests
-                                parallelRunnerEnvironments.Add(test.Id, testEnvironments);
-                            }
-                        }
-                    }
-
-                    // users can provide a custom report path which is the report base directory for all valid tests
-                    string reportPath = null;
-                    if (_ciParams.ContainsKey("fsReportPath"))
-                    {
-                        reportPath = _ciParams["fsReportPath"];
-                    }
-
-                    SummaryDataLogger summaryDataLogger = GetSummaryDataLogger();
-                    List<ScriptRTSModel> scriptRTSSet = GetScriptRtsSet();
-                    if (_ciParams.ContainsKey("fsUftRunMode"))
-                    {
-                        string uftRunMode = "Fast";
```

Comment:
Removing the default value `Fast` may cause the unknown behavior if `fsUftRunMode` key exists with **empty value**. Suggest to add the default value back, or run end-to-end test on such scenarios.

Reply:
The original code was:
                        string uftRunMode = "Fast";
                        uftRunMode = _ciParams["fsUftRunMode"];
The modified code is:
                        string uftRunMode = _ciParams["fsUftRunMode"];

I think you agree that in both cases we get the same result. 
The default value is always replaced by the value of _ciParams["fsUftRunMode"], even if it's empty.

Maybe the best solution to handle the unknown behavior is like this:
                        string uftRunMode = _ciParams["fsUftRunMode"];
                        if (string.IsNullOrEmpty(uftRunMode)) 
                        {
                            uftRunMode = "Fast";
                        }

Should we go with this last option?

Reply:
You're right, the value read from the `_ciParams` always overwrites the default one, regardless it is empty or not. So, I'm curious what will happen if `uftRunMode == string.Empty`?

If the empty string would cause the program crash, then I think your proposed solution is the best; otherwise, simply keep your changes as is - since the empty string can work there is not necessary to set default value which might cause the compatibility issue.